libtenzir/builtins/operators/where_map.cpp

```diff
@@ -277,83 +279,403 @@ auto make_where_map_function(function_plugin::invocation inv, session ctx,
-      const auto empty_type = type{fmt::to_string(mode), record_type{}};
+      const auto empty_type = type{"where", record_type{}};
       auto slice = table_slice{
         arrow::RecordBatch::Make(empty_type.to_arrow_schema(),
                                  list_values.length(), arrow::ArrayVector{}),
         empty_type,
       };
       slice = assign(args.capture, list_values, slice, ctx);
-      auto result = tenzir::eval(args.expr, slice, ctx);
-      TENZIR_ASSERT(not result.parts().empty());
-      if (result.parts().size() > 1) {
-        // TODO: We could do some attempt of unification here.
-        // TODO: The error message is bad. It's difficult to explain.
-        diagnostic::warning("expression type must not depend on the argument")
-          .primary(args.expr)
+      auto ms = tenzir::eval(args.expr, slice, ctx);
+      TENZIR_ASSERT(not ms.parts().empty());
+      auto result = std::vector<series>{};
+      auto offset = int64_t{0};
+      auto next_offset = int64_t{0};
+      for (auto& values : ms.parts()) {
+        offset = next_offset;
+        next_offset += values.length();
+        if (values.as<null_type>()) {
+          auto builder = series_builder{field.type};
+          for (auto i = int64_t{0}; i < field.length(); ++i) {
+            builder.list();
+          }
+          result.push_back(builder.finish_assert_one_array());
+          continue;
+        }
+        const auto predicate = values.as<bool_type>();
+        if (not predicate) {
+          diagnostic::warning("expected `bool`, but got `{}`",
+                              values.type.kind())
+            .primary(args.expr)
+            .emit(ctx);
+          result.push_back(series::null(field.type, field.length()));
+          continue;
+        }
+        if (predicate->array->true_count() == predicate->length()) {
+          result.push_back(field.slice(offset, values.length()));
+          continue;
+        }
+        auto predicate_gen = predicate->values();
+        auto builder = series_builder{field.type};
+        match(field_list->type.value_type(), [&]<concrete_type T>(const T&) {
+          for (auto&& list : field_list->values()) {
+            if (not list) {
+              builder.null();
+              continue;
+            }
+            auto list_builder = builder.list();
+            for (auto&& element : *list) {
+              auto should_filter = predicate_gen.next();
+              TENZIR_ASSERT(should_filter);
+              if (should_filter->value_or(false)) {
+                list_builder.data(as<view<type_to_data_t<T>>>(element));
+              }
+            }
+          }
+          // Check that we actually did iterate over all evaluated
+          TENZIR_ASSERT(not predicate_gen.next());
+        });
+        result.push_back(builder.finish_assert_one_array());
+      }
+      return multi_series{std::move(result)};
+    });
+  });
+}
+
+auto make_map_function(function_plugin::invocation inv,
```

Comment:
Nit: There's no reason for this to be in the same translation unit anymore, as there's no more shared code with `where`.